main/src/cgeo/geocaching/storage/DataStore.java

```diff
@@ -3408,6 +3411,86 @@ public static void saveLists(final Collection<Geocache> caches, final Set<Intege
         }
     }
 
+    public static Set<String> getStoredLocationsInList(final int listId) {
+
+        final Set<String> locations = new TreeSet<>(String::compareTo);
+
+        init();
+
+        try (Cursor cursor = database.rawQuery(PreparedStatement.GET_STORED_LOCATIONS_IN_LIST.query, new String[] {Integer.toString(listId)})) {
+
+            if (cursor.moveToFirst()) {
+                do {
+                    locations.add(cursor.getString(cursor.getColumnIndex("location")));
+                } while (cursor.moveToNext());
+            }
+
+        } catch (final SQLiteDoneException ignored) {
+            // Do nothing, it only means we have no information on the cache
+        } catch (final Exception e) {
+            Log.e("DataStore.getStoredLocationsInList", e);
+        }
+
+        return locations;
+    }
+
+    public static Set<String> getAllStoredLocations() {
+
+        final Set<String> locations = new TreeSet<>(String::compareTo);
+
+        init();
+
+        try (Cursor cursor = database.rawQuery(PreparedStatement.GET_ALL_STORED_LOCATIONS.query, new String[0])) {
+
+            if (cursor.moveToFirst()) {
+                do {
+                    locations.add(cursor.getString(cursor.getColumnIndex("location")));
+                } while (cursor.moveToNext());
+            }
+
+        } catch (final SQLiteDoneException ignored) {
+            // Do nothing, it only means we have no information on the cache
+        } catch (final Exception e) {
+            Log.e("DataStore.getAllStoredLocations", e);
+        }
+
+        return locations;
+    }
+
+    public static Set<String> getStoredLocationsForSet(Set<String> geocodes) {
```

Comment:
this should be able to utilize the DB for getting all locations for a set of geocodes, this way we can skip creating Geocache objects